scripts/openwebnet-discovery.nse

```diff
@@ -0,0 +1,199 @@
+who[14] = "Light+shutters actuators lock"
+who[15] = "WHO=15 - CEN"
+who[16] = "Sound System"
+who[17] = "Scenario Programming"
+who[18] = "Energy Management"
+who[24] = "Lighting Management"
+who[25] = "WHO=25 - CEN plus"
+who[1000] = "Diagnostic"
+who[1001] = "Automation Diagnostic"
+who[1004] = "Heating Diagnostic"
+who[1008] = "Door Entry System Diagnostic"
+who[1013] = "Device Diagnostic"
+
+-- Initiates a socket connection and returns the received data.
+local function get_socket(host, port, request)
+
+  local sd, response, early_resp = comm.opencon(host, port, request)
+
+  if sd == nil then
+    return nil, nil, "Socket connection error."
+  end
+
+  if not response then
+    return nil, nil, "Poor internet connection or no response."
+  end
+
+  if response == "*#*0##" then
+    return nil, nil, "Received a negative ACK as response."
+  end
+
+  stdnse.sleep(2)
+
+  -- Request for fetching Gateway address
+  sd:send(request)
+
+  local status, data = sd:receive_buf("*#*1##", false)
+  if data == "EOF" then
+    return nil, nil, "Received EOF with no response."
+  end
+
+  if data == "TIMEOUT" then
+    return nil, nil, "Timeout occurred."
+  end
+
+  -- Ignore if the response is EOF which is of length 3
+  -- Gateway length will be greater than 3
+  if #data > 3 then
+    -- Appending the delimiters to the data received for showing output
+    data = "*#*1##" .. data .. "1##"
+    return sd, data, nil
+  end
+end
+
+-- Removes *#*1## from the beginning and ending
+local function trim_begin_and_end(gateway)
+  local trim_begin = string.sub(gateway, 7)
+  local trim_end = string.sub(trim_begin, 1, -6)
+  return trim_end
+end
+
+-- Returns table after appending the delimiter
+-- The return table contains the list of devices
+local function custom_split(delimiter, resultant)
+  -- Trim at the end point to check for multiple entries
+  local fields = stdnse.strsplit(delimiter, resultant)
+
+  local results = {}
+  for _,v in pairs(fields) do
+    if #v > 0 then
+      table.insert(results, v .. delimiter)
+    end
+  end
+
+  return results
+end
+
+action = function(host, port)
+
+  local output = {}
+
+  local sd, gateway, err = get_socket(host, port, "*#13**15##")
+
+  -- Socket connection creation failed
+  if sd == nil then
+    return err
+  end
+
+  output["Gateway"] = gateway
+
+  local resultant = trim_begin_and_end(gateway)
+  local results = custom_split("##", resultant)
+
+  for _, v in pairs(results) do
+    -- Retrieving the device ID from Gateway
+    -- Its WHO value is equal to 13
+    local device_id = string.match(v,"(%d+)##$")
+    if device_id == nil then
+      -- Do nothing, ignore the failed case
+    elseif tonumber(device_id) > 0 then
+      output["Device"] = device[tonumber(device_id)]
+    end
+  end
+
+  -- Fetching list of each device
+  for _, v in pairs(who) do
+
+    stdnse.debug("Fetching the list of " .. v .. " devices.")
+
+    local sd, data, err = get_socket(host, port, "*##*#" .. _ .. "*0##")
```

Comment:
If I'm using the last socket we opened, I'm getting the previous response only even after sending the new data. That's the reason I'm creating new socket every time. I'm not sure of the reason behind this kind of response.